File: src/main/java/com/bike/shop/service/ClienteService.java
package com.bike.shop.service;

import com.bike.shop.dto.ClienteDTO;
import com.bike.shop.entity.Cliente;
import com.bike.shop.exception.DuplicateResourceException;
import com.bike.shop.exception.RecursoNoEncontradoException;
import com.bike.shop.repository.ClienteRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ClienteService {

    private final ClienteRepository clienteRepository;

    // GET todos los clientes
    public List<ClienteDTO> listarTodos() {
        return clienteRepository.findAll()
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    // GET buscar por documento
    public ClienteDTO buscarPorDocumento(String documento) {
        Cliente cliente = clienteRepository.findById(documento)
                .orElseThrow(() -> new RecursoNoEncontradoException("Cliente no encontrado: " + documento));
        return toDTO(cliente);
    }

    // GET buscar por término (nombre, documento, email)
    public List<ClienteDTO> buscarPorTermino(String termino) {
        return clienteRepository.buscarPorTermino(termino)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    // POST crear cliente
    public ClienteDTO crear(ClienteDTO dto) {
        if (clienteRepository.existsById(dto.getDocumento())) {
            throw new DuplicateResourceException("Ya existe un cliente con documento: " + dto.getDocumento());
        }
        Cliente cliente = toEntity(dto);
        return toDTO(clienteRepository.save(cliente));
    }

    // PUT actualizar cliente
    public ClienteDTO actualizar(String documento, ClienteDTO dto) {
        Cliente cliente = clienteRepository.findById(documento)
                .orElseThrow(() -> new RecursoNoEncontradoException(
                        "Cliente no encontrado: " + documento));
        cliente.setNombre(dto.getNombre());
        cliente.setTelefono(dto.getTelefono());
        cliente.setEmail(dto.getEmail());
        cliente.setDireccion(dto.getDireccion());
        return toDTO(clienteRepository.save(cliente));
    }

    // Convertir Entity → DTO
    private ClienteDTO toDTO(Cliente c) {
        return new ClienteDTO(
                c.getDocumento(),
                c.getNombre(),
                c.getTelefono(),
                c.getEmail(),
                c.getDireccion()
        );
    }

    // Convertir DTO → Entity
    private Cliente toEntity(ClienteDTO dto) {
        return new Cliente(
                dto.getDocumento(),
                dto.getNombre(),
                dto.getTelefono(),
                dto.getEmail(),
                dto.getDireccion()
        );
    }
}

File: frontend/src/app/core/services/cliente.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Cliente } from '../models/cliente.model';

@Injectable({ providedIn: 'root' })
export class ClienteService {
  private apiUrl = `${environment.apiUrl}/clientes`;

  constructor(private http: HttpClient) {}

  listar(): Observable<Cliente[]> {
    return this.http.get<Cliente[]>(this.apiUrl);
  }

  buscarPorDocumento(documento: string): Observable<Cliente> {
    return this.http.get<Cliente>(`${this.apiUrl}/${documento}`);
  }

  buscarPorTermino(termino: string): Observable<Cliente[]> {
    return this.http.get<Cliente[]>(`${this.apiUrl}/buscar?termino=${termino}`);
  }

  crear(cliente: Cliente): Observable<Cliente> {
    return this.http.post<Cliente>(this.apiUrl, cliente);
  }

  actualizar(documento: string, cliente: Cliente): Observable<Cliente> {
    return this.http.put<Cliente>(`${this.apiUrl}/${documento}`, cliente);
  }
}